File: stooge sort/stoogesort.cpp
/******************************************************
Program Name: stoogesort
Author: Christian Roccanova
Date: 4/12/2018
Description: reads data from file into an array and then sorts via stooge sort.  Sorted data is then output to stooge.out
******************************************************/

#include <iostream>
#include <fstream>
#include <string>

using std::cin;
using std::cout;
using std::endl;
using std::string;
using std::ifstream;
using std::ofstream;


void stoogeSort(int arr[], int front, int back)
{
	int temp;
	int third;
	int fore;
	int aft;

	//swaps first and last elements if first is larger
	if (arr[front] > arr[back])
	{
		temp = arr[front];
		arr[front] = arr[back];
		arr[back] = temp;
	}

	//recursively sorts remainder of the array in sets of 2/3
	if (back - front + 1 > 2)
	{
		third = (back - front + 1) / 3;
		fore = back - third;
		aft = front + third;

		//sorts first 2/3
		stoogeSort(arr, front, fore);

		//sorts back 2/3
		stoogeSort(arr, aft, back);

		//sorts first 2/3 again for confirmation
		stoogeSort(arr, front, fore);
	}

}

int main()
{
	int rowLen;
	int rowNum;
	int rowTotal = 0;
	int rowCount = 0;
	int numArr[100000];
	string rowString;
	ifstream dataFile;
	ofstream outFile("stooge.out");


	dataFile.open("data.txt");
	//counts number of rows in data file
	while (!dataFile.eof()) {
		getline(dataFile, rowString);
		rowTotal++;
	}

	dataFile.close();

	dataFile.open("data.txt");

	//cycles through each row individually
	for (int i = 0; i < rowTotal; i++)
	{

		dataFile >> rowLen;

		//fills initial array
		for (int j = 0; j < rowLen; j++)
		{
			dataFile >> numArr[j];
		}

		stoogeSort(numArr, 0, rowLen-1);

		//outputs sorted array to file
		for (int j = 0; j < rowLen; j++)
		{
			outFile << numArr[j] << " ";
		}


		outFile << endl;


	}
	outFile.close();
	dataFile.close();

	return 0;
}

File: stooge sort/stoogeSort.py
#Christian Roccanova
#CS 325
#HW 2 - stooge sort

import math

def stoogeSort(data, first, last):
    
    #compare first and last elements and swap if first is greater than last
    if data[first] > data[last]:
        temp = data[first]
        data[first] = data[last]
        data[last] = temp

    if last - first + 1 > 2:
        third = math.ceil((last - first) / 3) #size of a third of the array
        front = last - third #index of end of front 2/3rds
        back  = first + third #index of start of back 2/3rds
        
        #sort front 2/3rds
        stoogeSort(data, first, front)

        #sort back 2/3rds
        stoogeSort(data, back, last)

        #sort front 2/3rds again
        stoogeSort(data, first, front)
        
  

#main
# get data from file
dataFile = open("data.txt", "r")
lines = dataFile.readlines()
dataFile.close()
lineCount = len(lines)

#separate lines
for i in range(0, lineCount):
    sortData = []

    #separate line elements
    currLine = lines[i].split()    
    dataLength = int(currLine[0])
    
    #append line elements to data array
    for j in range(1, dataLength + 1):
        sortData.append(int(currLine[j]))

    #call stooge sort  
    n = len(sortData)
    stoogeSort(sortData, 0, n - 1)
   
    #write data to output file
    stoogeOut = open("stooge.out", "a")
    for k in range (0, dataLength):
        stoogeOut.write(str(sortData[k]))
        stoogeOut.write(" ")
    stoogeOut.write("\n")
    stoogeOut.close()
